test/python/synthesis/test_qsd_synthesis.py

```diff
@@ -0,0 +1,390 @@
+                    0.0 + 0.0j,
+                    0.0 + 0.0j,
+                    0.0 + 0.0j,
+                    0.0 + 0.0j,
+                ],
+            ]
+        )
+
+        gate = UnitaryGate(mat)
+        qc = QuantumCircuit(3)
+        qc.append(gate, range(3))
+        try:
+            qc.to_gate().control(1)
+        except UnboundLocalError as uerr:
+            self.fail(str(uerr))
+
+    def _create_random_multiplexed_gate(self, num_qubits):
+        num_blocks = 2
+        blocks = [scipy.stats.unitary_group.rvs(2 ** (num_qubits - 1)) for _ in range(num_blocks)]
+        mat = scipy.linalg.block_diag(*blocks)  # control on "top" qubit
+        multiplexed_gate = UnitaryGate(mat)
+        return multiplexed_gate, mat
+
+    def test_tensor_block_uc_2q(self):
+        """Create 2q gate with multiplexed controls"""
+        num_qubits = 2
+        gate, _ = self._create_random_multiplexed_gate(num_qubits)
+        for layout in itertools.permutations(range(num_qubits)):
+            # create gate with "control" on different qubits
+            qc = QuantumCircuit(num_qubits)
+            qc.append(gate, layout)
+            hidden_mat = Operator(qc).data
+            for j in range(num_qubits):
+                um00, um11, um01, um10 = qsd._extract_multiplex_blocks(hidden_mat, k=j)
+                # Check the off-diagonal
+                is_mult = qsd._off_diagonals_are_zero(um01, um10)
+                if is_mult:
+                    qc_uc = QuantumCircuit(num_qubits)
+                    ucgate = UCGate([um00, um11])
+                    qc_uc.append(ucgate, layout)
+                    uc_op = Operator(qc_uc)
+                    self.assertEqual(Operator(qc), uc_op)
+
+    def _get_multiplex_matrix(self, um00, um11, k):
+        """form matrix multiplexed wrt qubit k"""
+        halfdim = um00.shape[0]
+        dim = 2 * halfdim
+        ndim = halfdim.bit_length()
+        ure4 = np.zeros((2, halfdim, 2, halfdim), dtype=complex)
+        ure4[0, :, 0, :] = um00
+        ure4[1, :, 1, :] = um11
+        urend = ure4.reshape((2,) * ndim + (2,) * ndim)
+        urend = np.moveaxis(urend, ndim, k + ndim)
+        urend = np.moveaxis(urend, 0, k)
+        ure = urend.reshape(dim, dim)
+        return ure
+
+    def test_tensor_block_3q(self):
+        """Create 3q gate with multiplexed controls"""
+        num_qubits = 3
+        gate, _ = self._create_random_multiplexed_gate(num_qubits)
+        for layout in itertools.permutations(range(num_qubits)):
+            # create gate with "control" on different qubits
+            qc = QuantumCircuit(num_qubits)
+            qc.append(gate, layout)
+            hidden_mat = Operator(qc).data
+            for j in range(num_qubits):
+                um00, um11, um01, um10 = qsd._extract_multiplex_blocks(hidden_mat, k=j)
+                # Check the off-diagonal
+                is_mult = qsd._off_diagonals_are_zero(um01, um10)
+                if is_mult:
```

Comment:
Same comment as for the previous test: can we also check that `is_mult` is true at least once? Also, since `is_mult` is not used anywhere else, why not write ``if qsd._off_diagonals_are_zero(um01, um10) ...``?

Reply:
yes, that's a good idea!

Reply:
done [1597516](https://github.com/Qiskit/qiskit/pull/13688/commits/1597516bc45b4c126567b2ddc0d43608b7f884c2)